### ISSUE TITLE
Second LCD screen freezes

### ISSUE THREAD

**Issue by MobiFlight-Admin** [MobiFlight/MobiFlight-Connector#346]
## Problem summary ##
When using two lcd screens connected to one mega board with addresses 0x27 and 0x26, 
it works as expected when airplane is stationary.  The second screen (0x26) freezes and 
does not get updated as soon as the airplane starts to move.  The first screen (0x27) works
normally.  As the airplane lands and stops, the second screen starts to get updated again.  

## MF Version ##
8.0.2.9

## Steps to reproduce ##
1. Connect two lcd screens in parallel to i2c interface pins 20 and 21.
2. Configure two screens in MF Settings>Mobiflight Modules, addresses 0x27 and 0x26. 
   Upload configuration. "Mobiflight Rocks!" is displayed on both screens.
3. Create output configurations to display on the two screens.  In my case, I put flight
   data on first screen: heading, speed, altitude, vertical speed.  On second screen I put
   autopilot information: heading, speed, altitude, and vertical speed settings.
4. Start MSFS2020 and start a flight.

## Expected result ##
The two screens are updated in sync with the simulator throughout the flight.

## Actual result ##
When the airplane starts to speed up on the runway, the second screen freezes and is
not updated. The first screen continues to work.  When the airplane lands and stops, 
the second screen is again updated.

## Further details ##
This was tested with a fully loaded mega 2560 pro board (almost all pins connected).
Also tested offloading all pins to a second mega 2560 pro board, leaving only the two 
LCD screens connected to the mega board, with same results.
Tested connecting only the two LCD screens to a pro micro board with same results.
Only working solution so far is to connect each of the two screens to one board.
In my case, one mega board and one pro micro board.  

In this configuration, the system works well with both screens being updated and 
two or more screen pages can be switched between the two screens with one press of a 
button, which is not possible when using one board, but is another issue completely.

‌

[Issue created by jaime709: 2021-04-26]
[Last updated on bitbucket: 2021-05-28]

[Comment created by DocMoebiuz: 2021-05-28]
I think I found something in the code which might produce this problem. I will test and let you know.

**Comment by Mobiflight** [comment]
I found that I could fix this by using a different library but the library will use more memory and was causing a problem for the Micro FW and UNO FW to build.

I will have to investigate further.

**Comment by MobiFlight-Admin** [comment]
I can provide a workaround: duplicate the config for the 2nd display. It will update correctly then. I still haven't figured out why that is.

**Comment by JaimeLeon2** [comment]
> 
> 
> I guess it is more a timing problem, a pssible solution maybe not limited to the firmware. I would guess that after leaving the start position the values changes more often than in standstill. In the firmware the LCD is initialized with the standard frequency of 100kHz. The LCD needs 80Byte Data plus some overhead (didn't check exactly how much), means <90Byte (let's say 90Byte). As the LCD is operating in 4Bit mode a complete refresh needs 180Byte data to transfer. With 100kHz this takes ~18ms minus 3.2ms (32Byte I2C Buffer) = 15ms. One complete command to a LCD can be handled by the CMDMessenger (commandbuffer[96]) and is processed within. That mens that the CMDMessenger will be left after 15ms, but if the 2nd LCD will be updated directly after the 1st one the serialbuffer[64] gets overflowed (64Byte need 5,5ms to receive). The following command will not received completly and the CMDMessenger will ignore it and continue with the next complete command.
> 
> A first step would be to set the I2C speed to 400kHz (wire.setClock(400000)) within MFLCDDisplay.cpp during initialization). This would reduce the time to update the LCD to 4.5ms (data) minus 0.8ms (I2C buffer) = 3,7ms. That could be enough to get the LCD to work. But I might be wrong...
> 
> Another solution could be to implement a handshake in the communication. After processing a command the CMDMessenger sends an acknowledge (e.g."OK") and the connector waits for this command before the connector sends the next command. On the firmware side I don't see a big effort to implement this.
> 
> Maybe after more investigation other solutions are possible.

Sebastian provided me a few months back a firmware version running I2C at 400khz, but it didn't seem to have any positive effect on the issue at hand.

**Comment by elral** [comment]
I measured the time for writing 80 characters to the LCD, and even with 400kHz it is much more than I expected (~55ms).
It seems that more measures than switching to 400kHz are required.

**Comment by JaimeLeon2** [comment]
Another idea to reduce the traffic to the LCD screens. If the screens get refreshed by writing the entire 80 characters every time, why not instead just write the locations that change. 

IMHO this would effectively reduce the traffic by at least 75%. Granted that I probably don't know what I am talking, but it seems the LCD writing routine could be improved. In order to assemble the screen buffer, the locations of the changed data are already being determined, so couldn't the data be written to their proper location in the screen directly, instead of to the screen buffer?  Only the placeholder locations have "live" data.

**Comment by MobiFlight-Admin** [comment]
> Another idea to reduce the traffic to the LCD screens. If the screens get refreshed by writing the entire 80 characters every time, why not instead just write the locations that change.

It actually doesn't work well if you do it generically. In my setup I had 5 values set up:
HDG, IAS, VS, ALT which change frequently and the AP ALT, which doesn't change that much but i manipualte it with an encoder.

The 4 values that change frequently are distributed across 2 lines on a 4x20 LCD
```
AP-Mode: 
          (%%%%%)
IAS: $$$   #####:ALT
HDG: ^^^   &&&&&:VS
```

The experience is better if I send all 80 chars instead of the areas that change... it is the amount of messages that are sent immediately that dont work well. I would have to look for a throttling mechanism... which again would lead to a delay.

**Comment by MobiFlight-Admin** [comment]
more improvements with latest firmware.
